crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,262 @@
+# Kinds of types
+
+This document provides definitions for various kinds of types and classes,
+with examples of Python types and classes that satisfy these definitions.
+
+It doesn't attempt to be exhaustive.
+Definitions should only be added to this document if they are:
+
+1. Not covered by the ["Type system concepts" section of the Python typing spec](https://typing.readthedocs.io/en/latest/spec/concepts.html)
+1. Definitions that would be useful for developers working on red-knot.
+
+## Types
+
+A static Python type represents a set of possible values at runtime.
+The number of ways in which possible Python objects could be categorised into sets and subsets is infinite;
+thus, there is an infinite number of possible types in Python.
+
+The [typing spec](https://typing.readthedocs.io/en/latest/spec/) specifies various kinds of types
+that all Python type checkers should support,
+The desirable type checker behavior has been standardised to some degree for these specified types,
+but the spec is by no means exhaustive.
+There are more kinds of types in Python than are included in the spec,
+and many kinds of types that are left unspecified must be understood by a type checker
+in order for the type checker to accurately and precisely model Python's runtime semantics.
+
+Some types in Python have a known size.
+Most, however, have an infinite number of possible inhabitants and subtypes.
+For any type `T`, the union of all of the proper subtypes[^2] of `T` is exactly equal to `T`.
+
+## Singleton types
+
+A singleton type is a type for which it is known that there is
+(and can only ever be) exactly one known inhabitant of the type at runtime:
```

Comment:
nit: the use of the qualifier "known" in this sentence invites uncertainty about whether you are saying "there is definitely exactly one inhabitant, and it is known" or "there is exactly one known inhabitant, there may be other unknown ones."

You could be more explicit that it is the former, but I'm not sure it's worth it, vs just removing the word "known" -- I think it's obvious that if we can be sure there's exactly one inhabitant of type, that means we know what it is :)
```suggestion
(and can only ever be) exactly one inhabitant of the type at runtime:
```